rfcs/20221209-data-cleaning.md

```diff
@@ -0,0 +1,26 @@
+# RFC: Cleaning up the upstream database
+
+One of the steps of the [data refresh process for images][data-refresh] is cleaning data that is not fit for production. This process runs weekly as Airflow DAGs, the cleaned data is only saved to the API database (DB), which is replaced when the data refresh finishes, so it needs to be cleaned every time. This cleaning step takes up to 16 hours lately. It has been stable in duration given the strategy changed to perform the validations and transformations of the data upfront, when pulling from providers rather than when copying the upstream catalog DB into the API DB, but images ingested prior to this change remained untouched.
+
+The reason why the cleaning step is only applied to images and not to audio records is related to this change. The audio media type was integrated into Openverse after the creation of the MediaStore classes, which are the ones performing general and media-specific validations of the data pulled from providers.
+
+As seen, this is a costly process in terms of time and CPU resources, and it is inefficient to repeat it every time. Here I want to propose solving it in origin by doing a one-time pass by the image table of the upstream DB. We can follow the approach proposed by a previous Data Engineer contributor to [clean preexisting data using ImageStore][imagestore], see the details explained in the pull request and linked issue. The advantage of this way is that we apply not only the [cleaning steps defined in the ingestion server][cleanin-steps-is] but all the extra validations which were added later. I can foresee us using the same workflow in future opportunities if we require applying new modifications to the existing data, as long as its feasibility is confirmed.
+
+## Plan
+
+After evaluating what needs to be done to get this rid of the duplicated cleaning step, I collected the relevant issues of the openverse-catalog into a new milestone, [Data cleaning unification][milestone], and determined a rough plan to solve them:
+
+    1. Include the ingestion server cleaning steps in the ImageStore class
```

Comment:
Does that mean they are also happening in the `AudioStore` class or do we need to pull that logic into the `MediaStore` base class?

Reply:
From what I can see at the [`_enrich_tags`][enrich_tags_code] method (called from `clean_media_metadata`), it's only adding the provider from wich the tag is added. We still need to clean duplicates and complete the deny/exclude list.

So the placement of the URL validation is interesting. I thought it would fit better into the `ImageStore` class as we're concentrating there the validation. Do you think it should be moved early to the class or its current place is fine? 

[enrich_tags_code]: https://github.com/WordPress/openverse-catalog/blob/5933f712d2d017eb1d952c68fffc6f1606d58eb1/openverse_catalog/dags/common/storage/media.py#L250-L268

Reply:
> Does that mean they are also happening in the AudioStore class

As I understand it, the steps Olga mentions are already happening for both Audio and Image. `clean_media_metadata` is defined on the `MediaStore` and called by both subclasses, and both classes use the `UrlColumn` for their url fields (and thus get the url validation). Agreed that the duplication and denylist/accuracy threshold checks would need to be added, though. 

> So the placement of the URL validation is interesting. I thought it would fit better into the ImageStore class as we're concentrating there the validation.

I don't mind the url validation where it is, but you could certainly make a case that it should be moved -- but I would say to `clean_media_metadata` in the `MediaStore`, to make sure it continues to be used by Audio as well. In the past I've definitely taken a bit longer to find where that validation is happening, since it's happening in a different place!

Reply:
> Does that mean they are also happening in the AudioStore class or do we need to pull that logic into the MediaStore base class?

Sorry for being unclear: the meta_data clean up steps were extracted to the MediaStore when we added Audio. All of the Audio data fields also go through the clean up step in the Columns' `prepare_string` methods. So, in short, audio does not need to be cleaned. Except for one piece I'll mention below: the duplicate tags.

`_enrich_tags` method _does_ remove the tags that are on denylist: 
https://github.com/WordPress/openverse-catalog/blob/5933f712d2d017eb1d952c68fffc6f1606d58eb1/openverse_catalog/dags/common/storage/media.py#L267

But it doesn't remove the **duplicate tags**! So, all of the tags need to be de-duplicated. And we actually need to add tag de-duplication to the `_enrich_tags` method in `MediaStore`.

For the placement of URL validation: I like the clean separation of concerns where the column is responsible for validating and creating SQL insert/update strings. Would adding documentation to the `MediaStore` class about where the validation for various columns is happening help in discoverability? I think moving the validation to `ImageStore` would mean a lot more code duplication because we would be calling `validate` methods in both ImageStore and AudioStore, for all of the URL fields that we have.

I can see that it's unclear that a method called [`_create_tsv_row`](https://github.com/WordPress/openverse-catalog/blob/5933f712d2d017eb1d952c68fffc6f1606d58eb1/openverse_catalog/dags/common/storage/media.py#L185) actually also performs column validation. I wonder if renaming it and adding a clearer docstrings would help?